File: cpp/148_SortList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        int length = 1;
        ListNode* curr = head;
        while (curr) {
            curr = curr->next;
            length++;
        }
        return split(head, length);
        
    }
    
    ListNode* merge_sort(ListNode* a, ListNode* b) {
        if (!a || !b) return a ? a : b;
        else if (a->val > b->val) return merge_sort(b, a);

        a->next = merge_sort(a->next, b);
        return a;
    }
    
    ListNode* split(ListNode* head, int length) {
        if (!head || !head->next || length == 1) return head;
        
        int half = length / 2, i = 0;
        ListNode* prev = NULL;
        ListNode* curr = head;
        while (curr && i < half) {
            prev = curr;
            curr = curr->next;
            i++;
        }
        
        if (prev) prev->next = NULL;
        head = split(head, half), curr = split(curr, half % 2 == 1 ? half + 1 : half);
        return merge_sort(head, curr);
    }
    
};

File: python/148_SortList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        # find length of original list
        length, curr = 1, head
        while curr:
            length += 1
            curr = curr.next
        
        return self.split(head, length)
    
    def mergeSort(self, a, b):
        if not (a and b):
            return a or b
        elif a.val > b.val:
            return self.mergeSort(b, a);
        
        a.next = self.mergeSort(a.next, b)
        return a
            
    def split(self, head, length):
        half = length // 2
        right_half = half + half % 2
        if not (head and head.next) or half == 0: return head
        
        prev, curr, i = None, head, 0
        while curr and i < half:
            prev, curr = curr, curr.next
            i += 1
        
        if prev: prev.next = None
        head, curr = self.split(head, half), self.split(curr, right_half)
        return self.mergeSort(head, curr)
